specs/interop/superchain-weth.md

```diff
@@ -0,0 +1,161 @@
+amount of protocol code required to support native ETH. This generally points towards an app-layer solution if possible
+but does not preclude a protocol-layer solution as long as we minimize implementation size.
+
+### We want to maximize functionality
+
+Any solution should strive to maximize the functionality of the underlying message-passing protocol. Support for native
+ETH should not unnecessarily restrict the interop protocol.
+
+### Streamlining the user experience
+
+An ideal solution would make it possible to streamline the user experience such that ETH transfers between chains look
+just like ETH transfers between addresses on the same chain. Much of the user experience can be solved at the wallet
+level but it’s also important that the underlying protocol doesn’t require an unnecessary number of transactions to
+transfer ETH.
+
+## Constants
+
+| Name                     | Value                                        |
+| ------------------------ | -------------------------------------------- |
+| `SuperchainWETH` Address | `0x4200000000000000000000000000000000000024` |
+| `ETHLiquidity` Address   | `0x4200000000000000000000000000000000000025` |
+
+## SuperchainWETH
+
+### Invariants
+
+TODO
+
+### Implementation Sketch
+
+```solidity
+contract SuperchainWETH is WETH98, ISuperchainERC20 {
+  L1Block internal l1Block = L1Block(Predeploys.L1_BLOCK);
+  L2ToL2CrossDomainMessenger messenger = L2ToL2CrossDomainMessenger(Predeploys.L2_TO_L2_CROSSDOMAIN_MESSENGER);
+  ETHLiquidity liquidity = ETHLiquidity(Predeploys.ETH_LIQUIDITY);
+
+  function deposit() public payable override {
+    if (l1Block.isCustomGasToken()) revert IsCustomGasToken();
+    super.deposit();
+  }
+
+  function withdraw(uint256 wad) public override {
+    if (l1Block.isCustomGasToken()) revert IsCustomGasToken();
+    super.withdraw(wad);
+  }
+
+  function sendETH(uint256 chainId) external payable {
+    deposit();
+    _sendWETH(msg.sender, msg.value, chainId, true);
+  }
+
+  function sendETHTo(address to, uint256 chainId) external payable {
+    deposit();
+    _sendWETH(to, msg.value, chainId, true);
+  }
+
+  function sendERC20(uint256 wad, uint256 chainId) external {
+    _sendWETH(msg.sender, wad, chainId, true);
+  }
+
+  function sendERC20To(address to, uint256 wad, uint256 chainId) external {
+    _sendWETH(to, wad, chainId, true);
+  }
+
+  function finalizeSendWETH(address to, uint256 wad, bool unwrap) external {
+    if (msg.sender != address(messenger)) revert Unauthorized();
+    if (messenger.crossDomainMessageSender() != address(this)) revert Unauthorized();
+
+    // Mint the required liquidity.
+    if (l1Block.isCustomGasToken()) {
+      liquidity.mint(wad);
+    }
+
+    // Mint the WETH balance.
+    balanceOf[to] += wad;
+    emit Bridge(to, wad)
+
+    // If user wants to unwrap, try unwrapping.
+    if (unwrap) {
+      try this.unwrap(to, wad) {
```

Comment:
I think it'll be less error prone to put this in an if/else rather than a try catch

Reply:
Problem is if unwrap reverts for any reason then the whole thing fails and the tokens get lost.